### PR TITLE
Added Collection One Shot Uploader

### PR DESCRIPTION
**Pull request by bmbouter**
The import_collection task was used by the Galaxy API to receive an
artifact from mazer, but this reuses it for a new viewset in the Django
API.

https://pulp.plan.io/issues/5050
closes #5050

### CODE REVIEW

File: pulp_ansible/app/serializers.py [hunk 1]
```diff
@@ -75,6 +77,22 @@ class Meta:
...
+
+    file = serializers.FileField(
+        help_text=_("The collection tarball."),
+        required=True,
+    )
+
+    sha256 = serializers.CharField(
+        required=False,
+        default=None,
+    )
```
Comment (daviddavis) [inline pulp_ansible/app/serializers.py]: What's this field for? Maybe some help text would help?

File: pulp_ansible/app/tasks/collections.py [hunk 2]
```diff
@@ -106,3 +107,32 @@ def nowhere(*args, **kwargs):
 
         collections = Collection.objects.filter(pk__in=collections_pks)
         new_version.add_content(collections)
+
+
+def import_collection(artifact_pk):
+    """
+    Create a collection from an uploaded artifact.
```
Comment (daviddavis) [inline pulp_ansible/app/tasks/collections.py]: s/collection/Collection/

File: pulp_ansible/app/serializers.py [hunk 3]
```diff
@@ -75,6 +77,22 @@ class Meta:
         model = CollectionRemote
 
 
+class CollectionOneShotSerializer(serializers.Serializer):
+    """
+    A serializer for the Collection One Shot Upload API.
+    """
+
+    file = serializers.FileField(
+        help_text=_("The collection tarball."),
```
Comment (daviddavis) [inline pulp_ansible/app/serializers.py]: s/collection/Collection/

File: pulp_ansible/app/serializers.py (hunk 1 above)
Comment (bmbouter) [reply pulp_ansible/app/serializers.py]: It's an optional sha256 digest of the tarball being submitted, and if submitted will validate. I will add help-text in the next push, that is a good idea.

File: pulp_ansible/app/tasks/collections.py (hunk 2 above)
Comment (bmbouter) [reply pulp_ansible/app/tasks/collections.py]: will fix in next push, ty

File: pulp_ansible/app/serializers.py (hunk 3 above)
Comment (bmbouter) [reply pulp_ansible/app/serializers.py]: will fix in next push ty

File: pulp_ansible/app/urls.py [hunk 4]
```diff
@@ -10,9 +10,12 @@
     RoleVersionList
 )
 
+from pulp_ansible.app.viewsets import CollectionUploadViewSet
+
 
 urlpatterns = [
     url(r'pulp_ansible/galaxy/(?P<path>.+)/api/$', GalaxyVersionView.as_view()),
+    url(r'ansible/collection/$', CollectionUploadViewSet.as_view({'post': 'create'})),
```
Comment (daviddavis) [inline pulp_ansible/app/urls.py]: Why `collection` instead of `collections`?

**Comment by daviddavis** [comment]
This produces a 500 error. I would expect it to produce a 4xx or something similar:

```
echo 'blah' > collection.tgz
http --form POST :24817/ansible/collection/ file@collection.tgz sha256=abc123
```

**Comment by daviddavis** [comment]
I tested with a valid collection and sha256 and it worked great :+1:

File: pulp_ansible/app/urls.py (hunk 4 above)
Comment (bmbouter) [reply pulp_ansible/app/urls.py]: I think plural would be better. I'll switch it in next push.

**Comment by bmbouter** [comment]
I'll fix the 500 error also.

**Comment by bmbouter** [comment]
Now the 500 error is handled. This was a good idea.

```
(pulp) [vagrant@pulp3-source-fedora29 ~]$ http --form POST :24817/ansible/collections/ file@collection.tgz sha256=abc123
HTTP/1.1 400 Bad Request
Allow: POST, OPTIONS
Connection: close
Content-Length: 77
Content-Type: application/json
Date: Wed, 03 Jul 2019 16:43:49 GMT
Server: gunicorn/19.9.0
Vary: Accept, Cookie
X-Frame-Options: SAMEORIGIN

[
    "The provided sha256 value does not match the sha256 of the uploaded file."
]
```

File: CHANGES/5050.feature [hunk 5]
```diff
@@ -0,0 +1,2 @@
+A new Collection uploader has been added to the pulp_ansible API at
+``/pulp/api/v3/ansible/collection/``.
```
Comment (daviddavis) [inline CHANGES/5050.feature]: nitpick - needs to be collections

Comment (bmbouter) [reply CHANGES/5050.feature]: that's not a nitpick that's just correctness :)

**Pull request merged**